File: Backend-main/src/entities/DonationEntity.ts
import {
  MemorialContributionStatus,
  MemorialDonationWreath,
} from 'src/memorial/interface/memorial.interface';
import { transformInt } from 'src/util/transformer';
import {
  Entity,
  Column,
  CreateDateColumn,
  PrimaryGeneratedColumn,
  ManyToOne,
  JoinColumn,
  VirtualColumn,
  OneToMany,
} from 'typeorm';
import { UserEntity } from './UserEntity';
import { MemorialEntity } from './MemorialEntity';
import { DonationLikeEntity } from './DonationLikeEntity';

@Entity('donation')
export class DonationEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ type: 'uuid', name: 'memorial_id' })
  memorialId: string;

  @ManyToOne(() => MemorialEntity)
  @JoinColumn({ name: 'memorial_id' })
  memorial: MemorialEntity;

  @Column({ type: 'uuid', name: 'owner_id' })
  ownerId: string;

  @ManyToOne(() => UserEntity)
  @JoinColumn({ name: 'owner_id' })
  owner: UserEntity;

  @OneToMany(() => DonationLikeEntity, (like) => like.donation)
  likes: DonationLikeEntity[];

  @CreateDateColumn({ name: 'created_at', type: 'timestamptz', default: () => 'now()' })
  createdAt: Date;

  @Column({ type: 'int8', name: 'item_count', transformer: transformInt })
  itemCount: number;

  @Column({ type: 'int8', name: 'value_in_cents', transformer: transformInt })
  valueInCents: number;

  @Column({
    type: 'enum',
    name: 'wreath',
    enum: MemorialDonationWreath,
    enumName: 'memorial_wreath',
  })
  wreath: MemorialDonationWreath;

  @Column({
    type: 'enum',
    name: 'status',
    default: MemorialContributionStatus.DRAFT,
    enumName: 'memorial_contribution_status',
    enum: MemorialContributionStatus,
  })
  status: MemorialContributionStatus;

  @Column({
    type: 'text',
    name: 'checkout_session_id',
    nullable: true,
    default: null,
  })
  checkoutSessionId: string | null;

  @VirtualColumn({
    type: 'int',
    transformer: transformInt,
    query: (alias) => `(SELECT COUNT(*) FROM donation_like dl WHERE dl.donation_id = ${alias}.id)`,
  })
  totalLikes: number;
}

File: Backend-main/src/entities/DonationEntity.java
package entities;

import entities.enums.MemorialContributionStatus;
import entities.enums.MemorialDonationWreath;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Formula;

@Entity
@Table(name = "donation")
public class DonationEntity {
  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  @Column(name = "id", columnDefinition = "uuid")
  private UUID id;

  @Column(name = "memorial_id", columnDefinition = "uuid")
  private UUID memorialId;

  @ManyToOne
  @JoinColumn(name = "memorial_id")
  private MemorialEntity memorial;

  @Column(name = "owner_id", columnDefinition = "uuid")
  private UUID ownerId;

  @ManyToOne
  @JoinColumn(name = "owner_id")
  private UserEntity owner;

  @OneToMany(mappedBy = "donation")
  private List<DonationLikeEntity> likes;

  @CreationTimestamp
  @Column(name = "created_at", columnDefinition = "timestamptz")
  private OffsetDateTime createdAt;

  @Column(name = "item_count")
  private Long itemCount;

  @Column(name = "value_in_cents")
  private Long valueInCents;

  @Column(name = "wreath")
  private MemorialDonationWreath wreath;

  @Column(name = "status")
  private MemorialContributionStatus status = MemorialContributionStatus.DRAFT;

  @Column(name = "checkout_session_id")
  private String checkoutSessionId;

  @Formula("(SELECT COUNT(*) FROM donation_like dl WHERE dl.donation_id = id)")
  private Integer totalLikes;

  public UUID getId() {
    return id;
  }

  public void setId(UUID id) {
    this.id = id;
  }

  public UUID getMemorialId() {
    return memorialId;
  }

  public void setMemorialId(UUID memorialId) {
    this.memorialId = memorialId;
  }

  public MemorialEntity getMemorial() {
    return memorial;
  }

  public void setMemorial(MemorialEntity memorial) {
    this.memorial = memorial;
  }

  public UUID getOwnerId() {
    return ownerId;
  }

  public void setOwnerId(UUID ownerId) {
    this.ownerId = ownerId;
  }

  public UserEntity getOwner() {
    return owner;
  }

  public void setOwner(UserEntity owner) {
    this.owner = owner;
  }

  public List<DonationLikeEntity> getLikes() {
    return likes;
  }

  public void setLikes(List<DonationLikeEntity> likes) {
    this.likes = likes;
  }

  public OffsetDateTime getCreatedAt() {
    return createdAt;
  }

  public void setCreatedAt(OffsetDateTime createdAt) {
    this.createdAt = createdAt;
  }

  public Long getItemCount() {
    return itemCount;
  }

  public void setItemCount(Long itemCount) {
    this.itemCount = itemCount;
  }

  public Long getValueInCents() {
    return valueInCents;
  }

  public void setValueInCents(Long valueInCents) {
    this.valueInCents = valueInCents;
  }

  public MemorialDonationWreath getWreath() {
    return wreath;
  }

  public void setWreath(MemorialDonationWreath wreath) {
    this.wreath = wreath;
  }

  public MemorialContributionStatus getStatus() {
    return status;
  }

  public void setStatus(MemorialContributionStatus status) {
    this.status = status;
  }

  public String getCheckoutSessionId() {
    return checkoutSessionId;
  }

  public void setCheckoutSessionId(String checkoutSessionId) {
    this.checkoutSessionId = checkoutSessionId;
  }

  public Integer getTotalLikes() {
    return totalLikes;
  }
}
